Refactor auth and separate API from main app

tilgovi opened the pull request:
Provide a way to construct just the API as a standalone WSGI app
and include this wholesale as a view inside the main site. This
means that they can use separate authentication policies for the
following benefits:

- The API can be deployed and integrated into other applications
  completely independently.

- The API app can be configured with a simpler authentication policy
  that does not access the session.

To accomplish this:

- Provide a `create_api` function in `h.app` that can construct a WSGI
  application that serves only the API.

- Extract resource endpoint authorization into a tween. This tween sets
  the initial value of the `REMOTE_USER` environment key so that the
  provided API application use the `RemoteUserAuthenticationPolicy`.

- Integrate authentication with the API by having the callback for the
  authentication policy provide the OAuth client as a consumer role.
  By not accessing the OAuth attributes in the API views, enable the
  aforementioned integration flexibility.

- Use a normal `SessionAuthenticationPolicy` for the main site.

Isolate the Annotator Auth token handling to make it easy to migrate to
the new token system.

- Generate standard web token claims and some backwards compatibility
  claims for the Annotator Auth plugin.

- Use a tween to port the X-Annotator-Auth-Token headers to isolate the
  legacy code paths from the core authentication machinery.

The`h.oauth` package can disappear completely as everything is packed
neatly into `h.auth`.

Close #1296.

nickstenning commented on h/accounts/events.py:
```diff
@@ -5,6 +5,5 @@ def __init__(self, request, user):
 
 
 class LogoutEvent(object):
-    def __init__(self, request, user):
```
Not sure I understand why you've removed this. Surely it's useful to know which user just logged out?

nickstenning replied:
Although actually, I see that nothing is using it, so fair dos.

nickstenning commented on h/api.py:
```diff
@@ -204,16 +202,19 @@ def delete(context, request):
-        consumer = auth.Consumer(request.client.client_id)
-        return auth.User(userid, consumer, False)
+        for principal in request.effective_principals:
+            if principal.startswith('consumer:'):
```
I think this may be a transition issue, but while testing this locally I ran into the issue that `request.unauthenticated_userid` appears to be an integer, causing a `AttributeError` to be thrown by this line.

I'm not entirely sure where or when `userid` got set to an integer (presumably it should always be `user.username`) but we should probably handle this case.

nickstenning replied:
Looks like this is #1947, and not directly a result of this changeset.

tilgovi replied:
Rebased. We should be good now.

tilgovi commented:
Yep. You issue with getting an integer is due to horus calling `remember ` after the new registration event when autologin is on. This does not happen with it off.

The best fix would be to set that to false and if we want an equivalent to use our own setting and handle it like we are right now in the subscribers, just without horus mucking it up afterward.

This will be an issue in development but not production.

Separate PR.

nickstenning commented:
> The best fix would be to set that to false and if we want an equivalent to use our own setting and handle it like we are right now in the subscribers, just without horus mucking it up afterward.

Yep. Exactly what I've done in #1961.

tilgovi commented:
Any other issues?

nickstenning commented on h/auth.py:
```diff
@@ -0,0 +1,231 @@
+            request.client = client
+            request.user = request.session.get('userid')
+        elif request.client_secret is not None:
+            client_id = request.client_id
```
I'm not sure I understand when this branch would execute. Where (other than the tests) is `request.client_id` or `request.client_secret` set?

tilgovi replied:
pyramid_oauthlib creates these properties that reflect request parameters.
On Feb 19, 2015 10:45 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25012966>:
>
> > +class RequestValidator(_RequestValidator):
> > +
> > +    """Validates JSON Web Tokens."""
> > +
> > +    def authenticate_client(self, request):
> > +        if request.client_id is None:
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
>
> I'm not sure I understand when this branch would execute. Where (other
> than the tests) is request.client_id or request.client_secret set?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25012966>.
>

nickstenning replied:
Ah, right. For my taste that's a bit too much magic just to save you from typing `.params.get(...)`, but hey.

tilgovi replied:
Yeah. I'm not in love with it. But the fact is that oauthlib has its own
request object and this allows us to build OAuth components that are not
pyramid specific.

Essentially what pyramid_oauthlib does is provide request properties so
that the pyramid request object can round trip through oauthlib components
rather than having the overhead of creating a separate request object. It
also means that in places where the integration is tricky, like getting
clients and sessions, we have access to the registry from the validator
callbacks. Beyond that, it just provides a way to use the registry to
provide implementations of grant types, tokens, etc.

So, it's not meant to be magical except where we need pyramid stuff. It's
supposed to quack like an oauthlib request object, which realizes all
parameters as attributes.
On Feb 19, 2015 11:06 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25014820>:
>
> > +class RequestValidator(_RequestValidator):
> > +
> > +    """Validates JSON Web Tokens."""
> > +
> > +    def authenticate_client(self, request):
> > +        if request.client_id is None:
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
>
> Ah, right. For my taste that's a bit too much magic just to save you from
> typing .params.get(...), but hey.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25014820>.
>

nickstenning commented on h/auth.py:
```diff
@@ -0,0 +1,231 @@
+            request.client = client
+            request.user = None
+
+        return request.client is not None
```
Doesn't this imply that if `request.client` is already set (perhaps by accident) this would return `True`? If that's intended behaviour then this is a bit too clever for me. I'd much rather see an early return at the top of the function.

If it's *not* intended behaviour, then I'd rather us checking a local variable: `client is not None`.

tilgovi replied:
No. It's set explicitly in both branches.
On Feb 19, 2015 10:48 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25013170>:
>
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
> > +            client_secret = request.client_secret
> > +            client = get_client(request, client_id, client_secret)
> > +            request.client = client
> > +            request.user = None
> > +
> > +        return request.client is not None
>
> Doesn't this imply that if request.client is already set (perhaps by
> accident) this would return True? If that's intended behaviour then this
> is a bit too clever for me. I'd much rather see an early return at the top
> of the function.
>
> If it's *not* intended behaviour, then I'd rather us checking a local
> variable: client is not None.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25013170>.
>

nickstenning replied:
But "both branches" is not exhaustive. If `request.client_id is not None and request.client_secret is None` then we'll reach this return statement without visiting either branch.

tilgovi replied:
Ah, right. Will fix. Good catch.
On Feb 19, 2015 11:02 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25014447>:
>
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
> > +            client_secret = request.client_secret
> > +            client = get_client(request, client_id, client_secret)
> > +            request.client = client
> > +            request.user = None
> > +
> > +        return request.client is not None
>
> But "both branches" is not exhaustive. If request.client_id is not None
> and request.client_secret is None then we'll reach this return statement
> without visiting either branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25014447>.
>

nickstenning commented on h/auth.py:
```diff
@@ -0,0 +1,231 @@
+    """
+    groups = []
+
+    if getattr(request, 'client', None) is not None:
```
`getattr(a, b, None) == getattr(a, b)`

nickstenning replied:
What rubbish. I'm an idiot. Sorry for the noise.

nickstenning commented on h/models.py:
```diff
@@ -178,10 +178,11 @@ def get_mapping(cls):
-    """A basic implementation of :class:`h.oauth.IClient`."""
+    """A basic implementation of :class:`h.interfaces.IClient`."""
 
     client_id = None
```
These lines are redundant now.

nickstenning commented:
Other than the minor niggles raised above, this looks good. Just to check I understand what's going on here:

- the correct way of working out if a user is authenticated for the current request is now to inspect `request.effective_principals`
- for most of the application, this is filled by a (cookie-keyed) session authentication policy
- for the API, this is set by the pyramid `REMOTE_USER` authentication policy, which reads the CGI environment variable set by the `set_user_from_oauth` subscriber hook, which in turn uses the oauthlib integration to parse the bearer token out of the request
- for backwards compatibility, `X-Annotator-Auth-Token` is translated into the appropriate `Authorization: Bearer ...` header in a tween

tilgovi commented:
Got it completely right.
On Feb 19, 2015 11:21 AM, "Nick Stenning" <notifications@github.com> wrote:

> Other than the minor niggles raised above, this looks good. Just to check
> I understand what's going on here:
>
>    - the correct way of working out if a user is authenticated for the
>    current request is now to inspect request.effective_principals
>    - for most of the application, this is filled by a (cookie-keyed)
>    session authentication policy
>    - for the API, this is set by the pyramid REMOTE_USER authentication
>    policy, which reads the CGI environment variable set by the
>    set_user_from_oauth subscriber hook, which in turn uses the oauthlib
>    integration to parse the bearer token out of the request
>    - for backwards compatibility, X-Annotator-Auth-Token is translated
>    into the appropriate Authorization: Bearer ... header in a tween
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951#issuecomment-75117465>.
>

tilgovi commented:
Alright. I cleaned up the `authenticate_client` call to have a local `user` and `client` both initialized to `None`. I added some some assertions to the tests that `request.client` and `request.user` are the right things (`None` where appropriate, otherwise the client or user).

tilgovi commented:
> the correct way of working out if a user is authenticated for the current request is now to inspect request.effective_principals

Just to comment on this point you made, @nickstenning:

`request.effective_principals` is _the_ way to get the effective principals for a request in Pyramid.
`request.authenticated_userid` is _the_ way to get the verified user identifier.
`request.unaunthenticated_userid` is _the_ way to get the _unverified_ user identifier. It's often the same as `authenticated_userid` but sometimes it sidesteps some work.

Both `authenticated_userid` and `effective_principals` validate the `unauthenticated_userid`. They perform the same function, except the former only returns the "primary" principal. This is why I use `unauthenticated_userid` in `api.get_user`: I call `effective_principals` right after that. Maybe that's unnecessarily tricky. Actually, yes. I'll just change that. No trickiness. Leave it to the policy (or db session, etc) to cache results if it must.

But the pedantic point I wanted to make was that these are the functions Pyramid's view predicates use. So the way of "working out if a user is authenticated for the current request" is really more likely to look like securing views with permissions. The authorization policy is responsible for deciding which permissions the request has based on the traversal context and the effective principals.

nickstenning commented:
LGTM.
